### ISSUE TITLE
Remove ALEMBIC_LIB_USES_{BOOST,TR1} and use the std versions.

### ISSUE THREAD

**Issue by savoury1** [alembic/alembic#312]
Build of 1.7.15 fails on Linux (Ubuntu) due 97f1b3cc11c5881f7412687d0aaa1b24ee98da2f with these errors and a few similar (using Clang 10.0.0):

```
[ 11%] Building CXX object lib/Alembic/CMakeFiles/Alembic.dir/AbcCoreOgawa/OrData.cpp.o
cd /build/alembic-io-1.7.15/obj-x86_64-linux-gnu/lib/Alembic && /usr/bin/clang++-10  -DALEMBIC_EXPORTS -DPLATFORM=LINUX -DPLATFORM_LINUX -I/build/alembic-io-1.7.15/lib -I/build/alembic-io-1.7.15/obj-x86_64-linux-gnu/lib -I/usr/include/OpenEXR -I/usr/include/hdf5/serial  -g -O2 -fdebug-prefix-map=/build/alembic-io-1.7.15=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -DH5_USE_18_API -fPIC   -fPIC -o CMakeFiles/Alembic.dir/AbcCoreOgawa/OrData.cpp.o -c /build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/OrData.cpp
/build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/OrData.cpp:68:26: error: no matching conversion for functional-style cast from 'Alembic::AbcCoreOgawa::v12::OrData::Child *' to 'Alembic::Util::unique_ptr<Child []>'
            m_children = Alembic::Util::unique_ptr< Child[] > (
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/alembic-io-1.7.15/lib/Alembic/Util/Foundation.h:161:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'Alembic::AbcCoreOgawa::v12::OrData::Child *' to 'const Alembic::Util::v12::unique_ptr<Alembic::AbcCoreOgawa::v12::OrData::Child []>' for 1st argument
class unique_ptr : noncopyable
      ^
```

Build also fails with either GCC 7.5.0 or GCC 9.3.0 though with slightly different error messages for GCC than Clang:

```
[ 11%] Building CXX object lib/Alembic/CMakeFiles/Alembic.dir/AbcCoreOgawa/OrData.cpp.o
cd /build/alembic-io-1.7.15/obj-x86_64-linux-gnu/lib/Alembic && /usr/bin/c++  -DALEMBIC_EXPORTS -DPLATFORM=LINUX -DPLATFORM_LINUX -I/build/alembic-io-1.7.15/lib -I/build/alembic-io-1.7.15/obj-x86_64-linux-gnu/lib -I/usr/include/OpenEXR -I/usr/include/hdf5/serial  -g -O2 -fdebug-prefix-map=/build/alembic-io-1.7.15=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fvisibility=hidden -DH5_USE_18_API -fPIC   -fPIC -o CMakeFiles/Alembic.dir/AbcCoreOgawa/OrData.cpp.o -c /build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/OrData.cpp
/build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/OrData.cpp: In constructor 'Alembic::AbcCoreOgawa::v12::OrData::OrData(Alembic::Ogawa::v12::IGroupPtr, const string&, std::size_t, Alembic::AbcCoreAbstract::v12::ArchiveReader&, const std::vector<Alembic::AbcCoreAbstract::v12::MetaData>&)':
/build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/OrData.cpp:69:45: error: no matching function for call to 'Alembic::Util::v12::unique_ptr<Alembic::AbcCoreOgawa::v12::OrData::Child []>::unique_ptr(Alembic::AbcCoreOgawa::v12::OrData::Child*)'
                 new Child[ headers.size() ] );
                                             ^
In file included from /build/alembic-io-1.7.15/lib/Alembic/Util/All.h:40:0,
                 from /build/alembic-io-1.7.15/lib/Alembic/AbcCoreAbstract/Foundation.h:40,
                 from /build/alembic-io-1.7.15/lib/Alembic/AbcCoreAbstract/ArchiveReader.h:41,
                 from /build/alembic-io-1.7.15/lib/Alembic/AbcCoreAbstract/All.h:40,
                 from /build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/Foundation.h:40,
                 from /build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/OrData.h:40,
                 from /build/alembic-io-1.7.15/lib/Alembic/AbcCoreOgawa/OrData.cpp:37:
/build/alembic-io-1.7.15/lib/Alembic/Util/Foundation.h:169:5: note: candidate: Alembic::Util::v12::unique_ptr<T>::unique_ptr(T*) [with T = Alembic::AbcCoreOgawa::v12::OrData::Child []]
     unique_ptr( T* val ) : p(val)
     ^~~~~~~~~~
/build/alembic-io-1.7.15/lib/Alembic/Util/Foundation.h:169:5: note:   no known conversion for argument 1 from 'Alembic::AbcCoreOgawa::v12::OrData::Child*' to 'Alembic::AbcCoreOgawa::v12::OrData::Child (*)[]'
/build/alembic-io-1.7.15/lib/Alembic/Util/Foundation.h:164:5: note: candidate: Alembic::Util::v12::unique_ptr<T>::unique_ptr() [with T = Alembic::AbcCoreOgawa::v12::OrData::Child []]
     unique_ptr()
     ^~~~~~~~~~
/build/alembic-io-1.7.15/lib/Alembic/Util/Foundation.h:164:5: note:   candidate expects 0 arguments, 1 provided
```

Evidently the changes to AbcCoreOgawa/OrData.{h,cpp} cause issues with various compilers on Linux systems, as the build of Alembic 1.7.14 completed without error using any of the mentioned compilers under Ubuntu.

**Comment by lamiller0** [comment]
What flags are you passing into the cmake command?

I tried this with clang10 and clang11 and both built for me.  I'm noticing -std=c++11 showing up in my compile line that doesnt appear to show up in yours.

**Comment by savoury1** [comment]
That was the right question, as I didn't consider the common -std=c++11 flag which was indeed the cause of the build failure. Reason that flag was not appearing in my build is that I was passing ALEMBIC_LIB_USES_BOOST=ON to cmake. A test build for Ubuntu Focal is on [Launchpad](https://launchpad.net/~savoury1/+archive/ubuntu/build-tools-stage/+build/19953764) with a full build log if that helps at all. It's destined for a Blender [PPA](https://launchpad.net/~savoury1/+archive/ubuntu/blender) I've established recently.

So there's a conditional in CMakeLists.txt that skips adding -std=c++11 if Boost libraries are being used (version 1.71 was used for the test build). This then results in the build failure due the new code in 1.7.15 for the two mentioned files.

https://github.com/alembic/alembic/blob/a3ad9600f5b5299dffda3d285aeab89023d2d801/CMakeLists.txt#L122-L132 

The build works fine as soon as ALEMBIC_LIB_USES_BOOST=ON is removed from the cmake call. Seems Boost doesn't like the new unique_ptr code used. Is there any particular advantage to using Boost libraries for the Alembic build with a modern GCC or Clang? (my C++ knowledge is quite limited, so I'm really not sure of the pros and cons). If there's no advantage to using Boost libraries then I'll simply not use Boost for my builds of Alembic from here on.

**Comment by lamiller0** [comment]
ALEMBIC_LIB_USES_BOOST was provided because in the earlier days of Alembic, C++11 (and even tr1) wasn't supported on many compilers.  Since it mainly needs it for the shared_ptr, weak_ptr, and unique_ptr,  I don't believe there are any advantages to introducing a boost dependency here.

**Comment by savoury1** [comment]
OK, thanks for the clarification, and I will leave Boost out of the Alembic builds from here on given the C++11 standard is now well supported on newer compiler versions.

**Comment by waebbl** [comment]
Would it make sense to remove the `ALEMBIC_LIB_USES_{BOOST,TR1}` variables from CMakeLists.txt?

**Comment by lamiller0** [comment]
Yes it would, I would pair this change with some other dependency changes.

I'll change the title on this issue as a reminder for the future.

**Comment by lamiller0** [comment]
I've started this as part of this branch:

https://github.com/lamiller0/alembic/tree/python3_fixes

**Comment by lamiller0** [comment]
This has been merged into master as part of the 1.8.0 rc
